Fix k8s external IPs implementation

aanm opened the pull request:
```release-note
Fix implementation of k8s external IPs as described in kubernetes documentation
```

Fix: https://github.com/cilium/cilium/issues/9546
Fix: https://github.com/cilium/cilium/issues/5293

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9092)
<!-- Reviewable:end -->

aanm committed:
api/v1: add k8s ExternalIP type to api/v1

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/k8s: add deepcopy to k8s Service struct

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: disable testing for wrong implementation of k8s external IPs

Fixes: 46ae5ef2be6b ("pkg/k8s: add k8s external IPs support")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
docs: add documentation to use external IPs functionality

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme commented on bpf/lib/nodeport.h:
```diff
@@ -646,6 +634,30 @@ static inline int nodeport_lb4(struct __sk_buff *skb, __u32 src_identity)
 	}
 
+        cilium_dbg_lb(skb, DBG_GENERIC, 1, 2);
+        cilium_dbg_lb(skb, DBG_GENERIC, svc->reserved, svc->is_k8s_external_ip);
```
Left-over debugging?

jrajahalme commented on bpf/lib/nodeport.h:
```diff
@@ -646,6 +634,30 @@ static inline int nodeport_lb4(struct __sk_buff *skb, __u32 src_identity)
+
+	if (!svc->is_k8s_external_ip) {
+                cilium_dbg_lb(skb, DBG_GENERIC, 7, 7);
+                cilium_dbg_lb(skb, DBG_GENERIC, 0x79, 0x80);
```
Left-over debugging?

jrajahalme commented on bpf/lib/nodeport.h:
```diff
@@ -733,6 +745,9 @@ static inline int rev_nodeport_lb4(struct __sk_buff *skb, int *ifindex,
 			 &monitor);
 
+        cilium_dbg_lb(skb, DBG_LB4_LOOKUP_MASTER, ip4->daddr, ip4->protocol);
+        cilium_dbg_lb(skb, DBG_LB4_LOOKUP_MASTER, ip4->saddr, ip4->protocol);
```
Left-over debugging?

jrajahalme reviewed (approved):
Small nits about likely left-over debugging messages.

brb commented on bpf/lib/common.h:
```diff
@@ -553,7 +553,9 @@ struct lb6_service_v2 {
 	__u16 rev_nat_index;
 	__u16 weight;
-	__u16 pad;
+	__u8 is_k8s_external_ip:1, /* K8s External IPs */
```
`s/is_k8s_external_ip/external/`?

aanm replied:
Changed to `k8s_external` to distinguish to with possible external IPs (from a different cluster)

brb commented on bpf/bpf_sock.c:
```diff
@@ -204,12 +211,20 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
 	struct lb4_service_v2 *svc;
 	struct lb4_key_v2 key = {
 		.dport		= ctx_get_port(ctx),
+		.address        = ctx_get_ip4(ctx),
```
Setting `key.address` will make `127.0.0.1:$NODE_PORT` unavailable from a Pod.

aanm replied:
The `sock4_handle_node_port` will correctly populate `key.address`

brb commented on bpf/bpf_sock.c:
```diff
@@ -204,12 +211,20 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
 
+	svc = __lb4_lookup_service_v2(&key);
+	if (svc) {
+		if (unlikely(svc->is_k8s_external_ip)) {
```
I think a comment why we exclude ExternalIP SVC here might be useful for a reader.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -646,6 +631,23 @@ static inline int nodeport_lb4(struct __sk_buff *skb, __u32 src_identity)
 		return TC_ACT_OK;
 	}
 
+	if (!svc->is_k8s_external_ip) {
```
I think you don't want to skip here and you need to modify the if-statement, as this handles a remote backend reply, which means that the reverse SNAT translation should be done.

borkmann replied:
How does this set handle the case where you SNAT the external request to the remote backend and at the same time an app from the node itself directly connects to the same backend where kernel picks same source port. In other words I'm not seeing how you SNAT all other connections to avoid clashing tuples.

brb commented on cilium/cmd/service_list.go:
```diff
@@ -113,19 +115,19 @@ func printServiceList(w *tabwriter.Writer, list []*models.Service) {
-			str = fmt.Sprintf("%d\t%s\t\t",
-				service.ID, service.FrontendAddress)
+			str = fmt.Sprintf("%d\t%s\t%t\t\t",
+				service.ID, service.FrontendAddress, service.ExternalIP)
```
Does k8s allocate a ClusterIP for ExternalIP SVC? If not, then printing `FrontendAddress` for such service is not needed.

aanm replied:
Yes it does.

brb reviewed (changes requested):
I did the first-round of reviewing the code.

One additional suggestion: we should add an info log statement when a user runs with `--enable-node-port=false` saying that `Using kube-proxy implementation of ExternalIP Service`.

borkmann commented on cilium/cmd/bpf_lb_list.go:
```diff
@@ -69,22 +69,23 @@ func dumpSVC(serviceList map[string][]string) {
 				ip = "[::]"
 			}
-			entry = fmt.Sprintf("%s:%d (%d)", ip, 0, revNATID)
+			entry = fmt.Sprintf("%s:%d (%d) [FLAGS: 0x%x]", ip, 0, revNATID, flags)
```
I would prefer if we add actual human readable names instead of hex dump, eases debugging, also for the other places.

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -204,12 +211,20 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
 	if (!sock_proto_enabled(ctx))
 		return CONNECT_PROCEED;
 
+	svc = __lb4_lookup_service_v2(&key);
```
Should have a compile flag `ENABLE_EXTERNAL_SERVICES` so we don't pay the cost of looking up service twice here in fast-path if not used.

aanm committed:
Revert "Revert "test: add integration tests for k8s services with external IPs""

This reverts commit 02b80eeb10f1f60ae55f72396f864b74478557f4.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
Revert "Revert "test: wait for k8s external service in [kube|core]-dns""

This reverts commit f57f24ffbc0c313976b1d8c247d6241295535d06.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
add correct implementation of k8s externalIPs in datapath

If the user wants to check if the backend address is of externalIP type
he can check the flags field which will be set with `ExternalIPs` if the
service is type of external IP.

```
sudo cilium bpf lb list
SERVICE ADDRESS       BACKEND ADDRESS
192.0.2.233:82        10.16.56.85:80 (22)
                      0.0.0.0:0 (22) [FLAGS: ExternalIPs]
                      10.16.92.10:80 (22)
```

Fixes: 46ae5ef2be6b ("pkg/k8s: add k8s external IPs support")
Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on Documentation/gettingstarted/external-ips.rst:
```diff
@@ -0,0 +1,82 @@
+.. note::
+
+   ExternalIPs services depend on the :ref:`host-services` feature, therefore
+   a v4.19.57, v5.1.16, v5.2.0 or more recent Linux kernel is required.
```
It might be worth mentioning as well that v5.0 is not supported.

joestringer commented on pkg/maps/lbmap/ipv4.go:
```diff
@@ -222,6 +223,8 @@ func (s *Service4Value) GetCount() int        { return int(s.Count) }
 func (s *Service4Value) GetRevNat() int       { return int(s.RevNat) }
 func (s *Service4Value) RevNatKey() RevNatKey { return &RevNat4Key{s.RevNat} }
+func (s *Service4Value) SetFlags(flags uint8) { s.Flags = flags }
+func (s *Service4Value) GetFlags() uint8      { return s.Flags }
```
Could this return a `type ServiceFlags uint8` which implements `Stringer` to get human-readable flags?

joestringer reviewed (commented):
Minor nits on docs/cli side. Can review again if you'd like another pair of eyes on the main logic.

brb commented on bpf/lib/common.h:
```diff
@@ -561,7 +561,9 @@ struct lb6_service {
 	__u16 count;
 	__u16 rev_nat_index;
-	__u32 pad;
+	__u8 k8s_external:1, /* K8s External IPs */
```
Maybe s/k8s_external/external_ip/ ? Let's keep the datapath free from `k8s`.

aanm replied:
Discussed offline, keeping with k8s_external to distinguish between a "real" possible external IP nomenclature in the future.

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -214,6 +214,12 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
+		// a public IP address.
+		if (svc->k8s_external) {
+			return CONNECT_PROCEED;
+		}
```
No extra `{` `}` needed above. The UDP bits are missing below in `sock4_xlate_snd()`.

Is there a case where the externalIP service is overlapping with a nodeport service on that host? Afaik, it potentially could, so here we would hit the nodeport wildcard entry which seems fine. One thing that is a bit less clear to me is that depending on the order of insertion into the services map, we either override the IP:nodeport entry for the nodeport service with the external IP one or vice versa? Is this prevented from orchestration side somehow?

borkmann commented on bpf/lib/common.h:
```diff
@@ -561,7 +561,9 @@ struct lb6_service {
-	__u32 pad;
+	__u8 k8s_external:1, /* K8s External IPs */
+	     reserved:7;
+	__u8 pad[3];
```
Should be part of the v6 patch?

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -631,26 +631,26 @@ static inline int nodeport_lb4(struct __sk_buff *skb, __u32 src_identity)
-		return DROP_MISSED_TAIL_CALL;
-	}
-
-	ct_state_new.orig_dport = key.dport;
```
I tried to find where we use orig_dport in our code, but looks like the last remains on policy side where removed soe time ago, so we can actually get rid of ct_state's orig_dport in general.

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -631,26 +631,26 @@ static inline int nodeport_lb4(struct __sk_buff *skb, __u32 src_identity)
-		return TC_ACT_OK;
 	}
 
+	if ((svc == NULL) || (!svc->k8s_external)) {
```
Small nit: no extra `(` `)` for both needed.

borkmann replied:
What happens if this is overlapping with nodeport service, we'd never do the reverse NAT?

brb commented on Documentation/cmdref/cilium_service_update.md:
```diff
@@ -19,6 +19,7 @@ cilium service update [flags]
       --frontend string    Frontend address
   -h, --help               help for update
       --id uint            Identifier
+      --k8s-external       Set service as an external IP
```
`--external-ip`?

brb commented on Documentation/gettingstarted/external-ips.rst:
```diff
@@ -0,0 +1,83 @@
+.. _external-ips:
+
+*****************************
+Kubernetes externalIPs (beta)
```
Wouldn't it make sense to merge this doc with the NodePort GSG? Many parts seem to be identical between those two, so I'd create a section for the ExternalIP in the NodePort GSG.

brb commented on cilium/cmd/service_list.go:
```diff
@@ -60,10 +60,11 @@ func listServices(cmd *cobra.Command, args []string) {
 
 func printServiceList(w *tabwriter.Writer, list []*models.Service) {
-	fmt.Fprintln(w, "ID\tFrontend\tBackend\t")
+	fmt.Fprintln(w, "ID\tFrontend\tExternal IP\tBackend\t")
```
s/External IP/Service Type/, and then print a service type?

brb commented on pkg/k8s/service.go:
```diff
@@ -142,13 +143,52 @@ func ParseService(svc *types.Service) (ServiceID, *Service) {
 	}
 
-	return svcID, svcInfo
+	var svcInfoExternal []*Service
```
Why can't you just extend the `Service` struct by adding a field to store external IPs? Then in the `pkg/k8s/watchers/service.go` create the ExternalIP services in similar way as we currently do for NodePort? This would allow us to get rid of a lot of new code below.

aanm replied:
Sorry @brb I know we have briefly discuss this offline but I don't see a good way to do it properly since an externalIP is in the end a frontend. I'll be happy to discuss this again and address it before 1.7 is released

aanm reviewed (commented):
comment

aanm committed:
cilium/cmd: add cli functionality to create k8s external services

It might be useful to test the datapath without k8s running.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
add externalIPs implementation in Cilium

This commit adds the implementation for k8s externalIPs into the control
plane. The user will be able to observe the externalIPs services
implemented into Cilium by typing `cilium service list`:

```
$ cilium service list
ID   Frontend              Service Type   Backend
...
4    192.0.2.233:31389     ExternalIPs    1 => 10.16.48.207:80
                                          2 => 10.16.123.29:80
...
```

Contrary to what is currently implemented in kube-proxy, Cilium will not
perform any service translation for traffic made to an externalIP where
its origin comes from either the host or any pod running in locally in
that host. In kube-proxy the service translation for externalIPs will be
made for traffic originated from the pod but it will not be made for
traffic originated from the host.

Signed-off-by: André Martins <andre@cilium.io>

borkmann reviewed (approved):
Overall I'm okay with merging as-is, but I do have a few minor comments that would be great to do as a follow-up cleanup. Will comment separately.

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -210,10 +210,47 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
 	if (!sock_proto_enabled(ctx))
 		return CONNECT_PROCEED;
 
+#ifdef ENABLE_K8S_EXTERNAL_IP  /* ENABLE_K8S_EXTERNAL_IP */
```
Is there a way in the daemon where I'd have a --enable-external-ip flag to switch it on/off?

If required for nodeport, it could be enabled transparently, but it would at least give an option to opt-out if the additional overhead would not be needed and a way to debug potential nodeport/external IP separately.

Lets also add that flag to the doc then.

aanm replied:
Good point, will do this.

borkmann commented on Documentation/cmdref/cilium_service_update.md:
```diff
@@ -19,6 +19,7 @@ cilium service update [flags]
       --frontend string    Frontend address
   -h, --help               help for update
       --id uint            Identifier
+      --k8s-external       Set service as a k8s external IP
```
I don't think this nor the daemon flag should have any k8s mention in it, so lets do just `--external-ip`. Neither struct members in the BPF datapath since the concept itself is generic and could potentially be used outside of k8s w/o any problem. So I think it's better to remove all the k8s mentions in the datapath and agent/cli flags. We have this similarly with nodeport etc.

aanm replied:
@borkmann I've mentioned this in the previous comments, the mention of k8s is on purpose to distinguish from potential future work that involves a type of IP that is external to the cluster. Although the concept itself is generic the functionality is tightly couple to a k8s feature.

borkmann replied:
Ok, is there anything coming up from k8s you're aware of that would have this potential naming clash?

aanm replied:
Discussed offline

aanm committed:
cmd: add externalIPs flag

This allows to disable the k8s external IPs feature, enabled by default,
to debug potential bugs in nodeport/external IP separately.

Signed-off-by: André Martins <andre@cilium.io>
